File: data/scripts/pdst.py
#http://rosalind.info/problems/pdst/

def lettura(filename):
    data = []
    with open(filename) as f:
        record = ""
        first = True
        for riga in f:
            riga = riga.rstrip("\n")
            if riga[0] == ">":
                if not first:
                    data.append(record)
                record = ""
                first = False
            else:
                record += riga
    data.append(record)
    return data

def p_distance(s1, s2):
    diff = sum(1 for a, b in zip(s1, s2) if a != b)
    return diff / len(s1)

def main():
    seqs = lettura("dataset.txt")
    n = len(seqs)
    for i in range(n):
        riga = [p_distance(seqs[i], seqs[j]) for j in range(n)]
        print(" ".join(f"{v:.5f}" for v in riga))

if __name__ == "__main__":
    # execute only if run as a script
    main()

File: public/js-solutions/pdst.mjs
// Creating a Distance Matrix (Rosalind ID: PDST) - soluzione JavaScript
// indipendente, non una trascrizione di problem.py: stessa logica
// (parsing FASTA multi-record, poi matrice n x n delle p-distanze tra
// ogni coppia di sequenze, ciascuna formattata a 5 decimali), riscritta
// in modo idiomatico per JS.
//
// Contratto: riceve il contenuto testuale di dataset.txt, restituisce
// l'output testuale (stessa forma dell'output Python).
function parseFasta(testo) {
  const righe = testo.split("\n").map((r) => r.replace(/\r$/, ""));
  const data = [];
  let record = "";
  let first = true;
  for (const riga of righe) {
    if (riga.startsWith(">")) {
      if (!first) data.push(record);
      record = "";
      first = false;
    } else {
      record += riga;
    }
  }
  data.push(record);
  return data;
}

function pDistance(s1, s2) {
  let diff = 0;
  const n = Math.min(s1.length, s2.length);
  for (let i = 0; i < n; i++) {
    if (s1[i] !== s2[i]) diff++;
  }
  return diff / s1.length;
}

export default function solve(datasetText) {
  const seqs = parseFasta(datasetText);

  if (seqs.length === 0) {
    throw new Error("Input non valido: nessun record FASTA trovato");
  }

  const n = seqs.length;
  const righeOutput = [];
  for (let i = 0; i < n; i++) {
    const riga = [];
    for (let j = 0; j < n; j++) {
      riga.push(pDistance(seqs[i], seqs[j]).toFixed(5));
    }
    righeOutput.push(riga.join(" "));
  }

  return righeOutput.join("\n") + "\n";
}
